src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+pub unsafe fn test_sub(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestSub");
+    test_two_regions(base_ptr, count, timeout_checker, write_sub)
+}
+
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the result of multiplying a random value from the value read from the location
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_mul(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestMul");
+    test_two_regions(base_ptr, count, timeout_checker, write_mul)
+}
+
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the result of dividing a random value from the value read from the location
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_div(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestDiv");
+    test_two_regions(base_ptr, count, timeout_checker, write_div)
+}
+
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the OR result of a random value and the value read from the location
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_or(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestOr");
+    test_two_regions(base_ptr, count, timeout_checker, write_or)
+}
+
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the AND result of a random value and the value read from the location
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_and(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestAnd");
+    test_two_regions(base_ptr, count, timeout_checker, write_and)
+}
+
+/// Base function for `test_xor`, `test_sub`, `test_mul`, `test_div`, `test_or` and `test_and`
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// Write to each pair using the given `write_val` function
+/// Read and compare the two halves of the memory region
+unsafe fn test_two_regions<F>(
```

Comment:
To avoid having to start all your callback functions with an `unsafe {}` block, don't bother with the trait
and take a straight function pointer as an arg. That way, you can declare the function pointer `unsafe` and don't
have to repeat yourself in every function. Ex. `write_val: unsafe fn(*mut usize, *mut usize, usize)`.